etcd/src/main/scala/io/buoyant/etcd/Key.scala

```diff
@@ -0,0 +1,210 @@
+  }
+
+  def compareAndSwap(
+    value: Buf,
+    prevIndex: Option[Long] = None,
+    prevValue: Option[Buf] = None,
+    prevExist: Boolean = true
+  ): Future[NodeOp] = {
+    require(prevIndex.isDefined || prevValue.isDefined || !prevExist)
+
+    var params = noParams
+    for (i <- prevIndex) {
+      params = params :+ "prevIndex" -> i.toString
+    }
+    for (Buf.Utf8(v) <- prevValue) {
+      params = params :+ "prevValue" -> v
+    }
+    if (!prevExist) {
+      params = params :+ "prevExist" -> "false"
+    }
+    val Buf.Utf8(v) = value
+    params = params :+ "value" -> v
+
+    val req = mkReq(keysPrefixPath ++ key, Method.Put, params)
+    client(req) flatMap (toNodeOp(req, _, key, params))
+  }
+
+  def delete(
+    dir: Boolean = false,
+    recursive: Boolean = false
+  ): Future[NodeOp] = {
+    var params = noParams
+    if (dir) {
+      params = params :+ ("dir" -> "true")
+      if (recursive) {
+        params = params :+ ("recursive" -> "true")
+      }
+    }
+
+    val req = mkReq(keysPrefixPath ++ key, Method.Delete, params)
+    client(req) flatMap (toNodeOp(req, _, key, params))
+  }
+
+  private[this] def getIndex(node: Node): Long = node match {
+    case Data(_, idx, _, _, _) => idx
+    case Dir(_, idx, _, _, nodes) => nodes.foldLeft(idx)(_ max getIndex(_))
+  }
+
+  def events(
+    recursive: Boolean = false,
+    backoff: Stream[Duration] = Stream.empty
+  ): Event[Try[NodeOp]] = new Event[Try[NodeOp]] {
+    private[this] val origBackoff = backoff
+
+    def register(witness: Witness[Try[NodeOp]]) = {
+      @volatile var closing = false
+      @volatile var currentOp: Future[NodeOp] = Future.never
+
+      def loop(idx: Option[Long], backoff: Stream[Duration]): Unit =
+        if (!closing) {
+          val op = get(recursive, wait = idx.isDefined, waitIndex = idx)
+          currentOp = op
+          op respond {
+            case note@Return(op) =>
+              witness.notify(note)
+              loop(Some(getIndex(op.node) + 1), origBackoff)
+
+            case note@Throw(ApiError(ApiError.KeyNotFound, _, _, idx)) =>
+              witness.notify(note)
+              loop(Some(idx + 1), origBackoff)
+
+            case note@Throw(ApiError(ApiError.EventIndexCleared, _, _, idx)) =>
+              loop(Some(idx + 1), origBackoff)
+
+            case note@Throw(NonFatal(e)) =>
+              backoff match {
+                case wait #:: backoff =>
+                  witness.notify(note)
+                  loop(None, backoff)
+
+                case _ =>
+                  witness.notify(Throw(BackoffsExhausted(key, e)))
+              }
+
+            case note@Throw(_) =>
+              witness.notify(note)
+          }
+        }
+
+      loop(None, origBackoff)
+
+      Closable.make { _ =>
+        closing = true
+        currentOp.raise(new FutureCancelledException)
+        Future.Unit
+      }
+    }
+
+  }
+
+  def watch(backoff: Stream[Duration] = Stream.empty): Activity[NodeOp] =
```

Comment:
After going back on forth on this for a while, I think it would be needlessly restrictive to drop support for events---it exposes etcd's API with minimal interference.  I suspect that it makes sense to provide helpers over this API to model updating states of a single node, but I've removed the watch helper for now until we have a concrete use case for it (to determine whether it makes sense -- I suspect we care more about `def dataWatch: Activity[Option[Buf]]` and `def dirWatch(recursive: Boolean): Activity[Option[Dir]]`)